jsonschema-core.xml

```diff
@@ -419,13 +419,25 @@
                     <t>
                         A JSON Schema resource is a schema which is
                         <xref target="RFC6596">canonically</xref> identified by an
-                        <xref target="RFC3986">absolute URI</xref>.
+                        <xref target="RFC3986">absolute URI</xref>.  Schema resources MAY
+                        also be identified by URIs, including URIs with fragments,
+                        if the resulting secondary resource (as defined by
+                        <xref target="RFC3986">section 3.5 of RFC 3986</xref>) is identical
+                        to the primary resource.  This can occur with the empty fragment,
+                        or when one schema resource is embedded in another.  Any such URIs
+                        with fragments are considered to be non-canonical.
```

Comment:
You say...

> A URI with a fragment is neither canonical nor non-canonical

And yet this block says otherwise.

Reply:
I would suggest rephrasing too "a URI with a fragment is **not necessarily** canonical or non-canonical" -- canonical state is a boolean state (this isn't Schroedinger's schema!), but we can use neither "has a fragment" nor "doesn't have a fragment" as an indicator of canonical state.

It might be worth stating more clearly how to determine the canonical URI of any particular location in a schema:
- if there is an "$id" keyword at this level, then the canonical URI is the $id URI after being fully resolved against its base (and there is never a fragment)
- otherwise, the canonical URI is the canonical URI of the closest $id (or however you want to say "the innermost schema that contains this one, that has an "$id"), with a fragment that contains the json pointer from that $id location to here; OR, if there is no $id in the schema at all, then we use the base URI of the root schema itself, plus a fragment of the json pointer from the root.

Reply:
> but we can use neither "has a fragment" nor "doesn't have a fragment" as an indicator of canonical state

We can't? `$id` can have a fragment, but the resulting resolved URI will also then have a fragment, and this isn't allowed...

> If present, the value for this keyword MUST be a string, and MUST represent a valid URI-reference. This URI-reference SHOULD be normalized, and MUST resolve to an absolute-URI (without a fragment). Therefore, "$id" MUST NOT contain a non-empty fragment, and SHOULD NOT contain an empty fragment.

I've had this open for hours and I can't remember if I was going to say anything else.

Am I missing something?

Reply:
We're not talking just about `$id`s here, but canonical URIs everywhere.  e.g.:

```
{
  "$id": "https://myschema",
  "properties": {
    "foo": {
      "allOf": [
        { "type": "string" }
      ]
  }
}
```

The canonical URI of the subschema at `type" is "https://myschema#/properties/foo/allOf/0/type". It has a fragment, and this URI is canonical because it's using the base of the nearest `$id`. But we can't tell that it's canonical just from looking at the URI itself - we need to inspect the document it's in to determine that.

Reply:
I agree with all of the above you've said, however upon re-reading I do have one question...

> ...unless the fragment is a plain-name fragment, i.e. derived from an $anchor, because $anchor URIs are never canonical.

If `$anchor` producing URIs are non-canonical, does that mean the keyword doesn't need to be supported?

Obviously that's not the intent here, but that's how I read it logically.

Do we need to amend this section to include plain name fragments?

> An implementation MAY choose not to support addressing schema resource
>                         contents by URIs using a base other than the resource's canonical URI,
>                         plus a JSON Pointer fragment relative to that base.  Therefore, schema
>                         authors SHOULD NOT rely on such URIs, as using them may reduce interoperability.

https://github.com/json-schema-org/json-schema-spec/pull/1104/files#diff-bda51aab56b4b2e23fc9e7249e80fe6e262dcd552f35c5de69fd4a6bf3e96b50R1847-R1850

Reply:
Yes, that's a much nicer way of putting that.  The intent is surely just to say "you don't have to support non-canonical URIs that involve json pointers".

Personally I find handling these sorts of URIs to be no trouble at all, so I wouldn't mind if this entire paragraph went away, but I gather that others store schemas internally in a way that makes it difficult to use a json pointer that crosses an $id boundary.

Reply:
OK, I think that sounds like an agreement.

Reply:
> We're not talking just about $ids here, but canonical URIs everywhere. e.g. in:... - @karenetheridge 

I'm not sure that's correct. In re-reading every use of "canonical" in the spec, we either say "canonical URI" or "canonical URI plus pointer fragment".

Specifically, in unmodified (in terms of the patch branch) sections of the spec from 2020-12: https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.appendix.A

Reply:
> Yes, that's a much nicer way of putting that. - @karenetheridge 

To what were you referencing, if you can recall?

Reply:
> If `$anchor` producing URIs are non-canonical, does that mean the keyword doesn't need to be supported?
>
> Obviously that's not the intent here, but that's how I read it logically.
>
> Do we need to amend this section to include plain name fragments?
>
> >  An implementation MAY choose not to support addressing schema resource
    contents by URIs using a base other than the resource's canonical URI,
    plus a JSON Pointer fragment relative to that base. Therefore, schema
    authors SHOULD NOT rely on such URIs, as using them may reduce interoperability.

Yes, I think this unintentionally excludes anchor fragments and a few other things that should be ok. I think that putting this in terms of the canonical URI is not correct. Here are a few examples to show what I mean.

```json
{
  "$id": "/root",
  "type": "object",
  "properties": {
    "aaa": { "$ref": "/embedded" },
    "bbb": { "$ref": "#/$defs/embedded" },
    "ccc": { "$ref": "/embedded#/$defs/abc" },
    "ddd": { "$ref": "/embedded#xxx" },
    "eee": { "$ref": "#/$defs/embedded/$defs/abc" },
    "fff": { "$ref": "#xxx" }
  },
  "$defs": {
    "embedded": {
      "$id": "/embedded",
      "type": "number",
      "$defs": {
        "abc": {
          "$anchor": "xxx",
          "type": "string"
        }
      }
    }
  }
}
```

The references at "aaa" and "ccc" represent what MUST be supported. "bbb" and "ddd" show non-canonical references that the clause above technically excludes from being required to support, but there's no reason not to support them. "eee" is the only case the above clause should be referring to. (I'm not sure why it's even allowed. It should be forbidden. But, that's off topic and nothing we can change in a patch release.) I'm not even sure what the spec technically says about "fff", but it should be invalid because there is no "xxx" anchor in the "/root" schema resource. It's in the "/embedded" schema resource.

So, I'd remove mention of canonical URIs from that clause and say something like, the result of following a URI with a fragment that points to a location outside the current schema resource is undefined. Then I'd get rid of all mention of canonical URIs because it's no longer necessary to describe any feature of JSON Schema.

Reply:
I don't THINK I dissagree.

"xxx" does not resolve for the reasons you've stated.

"eee" is the target of what isn't required to be supported.
I think people had a hard time conceptually with multiple URIs being valid, specifically across `$id` boundaries.

For every `$id` boundary you cross, you increase the number of potential resolveable URI locations by n^2.

I guess that could start to become a problem, but I haven't seen anyone raise any such issue.

Not being able to remember why this was done, I did some digging and found https://github.com/json-schema-org/json-schema-spec/commit/a93b605d2f5d72b0de824081567e435782c32ce2

Specifically...

> While URIs for subschemas with "$id" and their chidren can be
constructed using the base URI from a parent schema, these
URIs are non-canonical, and their behavior is undefined.  This is
on the grounds that switching between embedding and referencing
schema resources should behave essentially identically.

Essentially, taking @jdesrosiers above example, if we assume it's a bundled "compound schema" document, it works. However if the document was unbundled back into individual resources, "eee": `"$ref": "#/$defs/embedded/$defs/abc"` would no longer be resolveable.

This is because the reference at "eee" crosses resource boundary (aka a schema location with `$id` at the non-root location.

We called references such as "eee" "shadowed JSON Pointers" in discussions, it would seem.

Reply:
> > Yes, that's a much nicer way of putting that. - @karenetheridge
> 
> To what were you referencing, if you can recall?

The statement from you immediately above, where you propose new wording.

Reply:
Regarding plain name fragments, canonical uris, where they appear and what is needed to be supported \[copied from slack\]:

If the spec says (paraphrased) "you only MUST support references to canonical uris", and canonical uris can't have fragments, then it's impossible to have a reference to any location that doesn't have an $id. e.g. "$ref": "#/$defs/foo".  I think it's more correct to say that every position in the document has one and exactly one canonical URI that describes it, and zero or more other URIs that MAY also be used to refer to it (which some implementations may or may not choose to support), with the exception of plain-name fragments (generated via $anchor) which MUST also be supported.

Reply:
I agree that @karenetheridge's summary is what the spec is trying to say. I still think it's unnecessarily complicated, but that's something that would need to be addressed elsewhere.

> zero or more other URIs that MAY also be used to refer to it (which some implementations may or may not choose to support)

I dislike this phrasing. I think it's backwards. It makes it sound like this is an optional feature that some will choose to support and others will not. It's not a feature. It's always wrong. Just because it MAY happen to work in some implementations doesn't mean it's not always wrong.

It's like duplicate properties in JSON. Given, `{ "a": 1, "a": 2 }`, if I expect `/a` to be `2` and I get `2`, then the implementation appears to be working as I expect. I might think that this implementation "supports" the second property definition overriding the second. But that's not a real feature and not a reliable behavior.

The spec should say that the behavior of referencing a location using a non-canonical URI is undefined. Your schema might work the way you expected in your favorite implementation, but your schema is still wrong.

Reply:
> I dislike this phrasing. I think it's backwards. It makes it sound like this is an optional feature that some will choose to support and others will not. It's not a feature. It's always wrong.

It's not wrong. We even give examples of this usage in the spec: https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.appendix.A -- see "non-canonical URI with fragment relative to root.json"

The only reason to discourage the usage of non-canonical URIs is because they might break if the document is broken up into a different form, e.g. some parts bundled elsewhere.  There are some usecases where referencing locations relative to the root of the document, rather than relative to the nearest `$id`, might even be the most convenient implementation.

Reply:
> We even give examples of this usage in the spec

That explains what is considered canonical and what isn't. It's not an example of valid usage. It's there specifically to point out what isn't valid.

> The only reason to discourage the usage of non-canonical URIs is because they might break if the document is broken up into a different form, e.g. some parts bundled elsewhere.

Yeah, that's a pretty important reason.

> There are some usecases where referencing locations relative to the root of the document, rather than relative to the nearest `$id`, might even be the most convenient implementation.

I'm going to need proof for that assertion. I don't believe such a case exists.

Reply:
"pretty important" is subjective. Bundling can be completely irrelevant to some usecases, and quite useful to others. In any usecase where bundling is unimportant, non-canonical URIs can be valid and useful. Whether you believe in them or not is irrelevant.

> The schemas at the following URI-encoded JSON Pointers (relative to the root schema) have the following base URIs, **and are identifiable by any listed URI in accordance with sections 5 and 9.2.1 above**. 

Section 9.2.1 says:

> Since URIs involving JSON Pointer fragments relative to the parent schema resource's URI cease to be valid when the embedded schema is moved to a separate document and referenced, applications and schemas SHOULD NOT use such URIs to identify embedded schema resources or locations within them. 

Note: "should not", not "must not".

Why is this still being discussed? If you want to make this usecase prohibited, submit a proposal for the next draft.

Reply:
> Why is this still being discussed? If you want to make this usecase prohibited, submit a proposal for the next draft.

I did veer off course a little. I'm sorry about that and the confusion that caused. But right now, I'm not arguing to prohibit anything, nor am I arguing to change anything the spec says. I'm saying that the behavior of referencing a location using a non-canonical URI is undefined, not that it's prohibited. You can do it, but there's no guarantee what behavior you will get. The spec would still describe the same things, it's just framed in different way.

Perhaps the cause of the confusion is my use of the word "wrong". When I say "wrong" I don't mean invalid. I mean it in the same way that duplicate properties in JSON is "wrong". It's not an invalid document, but it's conceptually a "wrong" way to use JSON. It's syntactically valid, but semantically nonsensical.

The words you used are more in line with the way the spec is currently written. Obviously I dislike the way it's framed, but if changing that is too much for this PR, that's fine. The scope of this PR was never to fix everything, it was just to try to make a few small improvements.

Reply:
I made this initial comment as I was confused about the seeming contradiciton between the first comment and this section in the PR, however after review, I no longer believe they are contradictory, and so I'll resolve it.